compliance/collection/nodeinventorizer/caching_inventorizer.go

```diff
@@ -0,0 +1,284 @@
+	cache := readInventoryWrap(c.inventoryCachePath)
+	if cache != nil && cache.Inventory != nil && cache.ValidUntil.After(time.Now()) {
+		log.Debugf("Using cached node scan (valid until %v)", cache.ValidUntil)
+		return cache.Inventory, nil
+	}
+
+	// check for existing backoff, wait for specified duration if needed, then persist the new backoff duration
+	backoffDuration := c.initialBackoff
+	if cache != nil {
+		backoffDuration = c.validateBackoff(cache.BackoffDuration)
+	}
+
+	if backoffDuration > c.initialBackoff {
+		log.Warnf("Found existing node scan backoff file - last scan may have failed. Waiting %v seconds before retrying", backoffDuration.Seconds())
+		c.backoffWaitCallback(backoffDuration)
+	}
+
+	// Write backoff duration to cache
+	backoff := inventoryWrap{BackoffDuration: c.calcNextBackoff(backoffDuration)}
+	if err := writeInventoryWrap(backoff, c.inventoryCachePath); err != nil {
+		log.Warnf("Error writing node scan backoff file: %v", err)
+	}
+
+	// if no inventory exists, or it is too old, collect a fresh one
+	newInventory, err := collectInventory(nodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Write inventory to cache
+	inventory := inventoryWrap{
+		ValidUntil:      time.Now().Add(c.cacheDuration),
+		BackoffDuration: 0,
+		Inventory:       newInventory,
+	}
+	if err := writeInventoryWrap(inventory, c.inventoryCachePath); err != nil {
+		return nil, errors.Wrap(err, "persisting inventory to cache")
+	}
+
+	return newInventory, nil
+}
+
+func (c *CachingScanner) calcNextBackoff(currentBackoff time.Duration) time.Duration {
+	nextBackoffInterval := currentBackoff * backoffMultiplier
+	if nextBackoffInterval > c.maxBackoff {
+		return c.maxBackoff
+	}
+	return nextBackoffInterval
+}
+
+// validateBackoff ensures that a given duration does not exceed the max backoff setting
+func (c *CachingScanner) validateBackoff(backoff time.Duration) time.Duration {
+	if backoff > c.maxBackoff {
+		return c.maxBackoff
+	}
+	return backoff
+}
+
+func readInventoryWrap(path string) *inventoryWrap {
+	cacheContents, err := os.ReadFile(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No node scan cache file found, will run a new scan")
+			return nil
+		}
+		log.Warnf("Unable to read node scan cache, will run a new scan. Error: %v", err)
+		return nil
+	}
+
+	// deserialize stored inventory
+	var wrap inventoryWrap
+	if err := json.Unmarshal(cacheContents, &wrap); err != nil {
+		log.Warnf("Unable to load node scan cache contents, will run a new scan. Error: %v", err)
+		return nil
+	}
+	return &wrap
+}
+
+func writeInventoryWrap(w inventoryWrap, path string) error {
+	jsonWrap, err := json.Marshal(&w)
+	if err != nil {
+		return err
+	}
+
+	if err := os.WriteFile(path, jsonWrap, 0600); err != nil {
+		return err
+	}
+	return nil
+}
+
+// collectInventory scans the current node and returns the results as storage.NodeInventory object
+func collectInventory(nodeName string) (*storage.NodeInventory, error) {
```

Comment:
I am leaning towards exporting this function into a separate object that would implement the `nodeInventorizer`  interface and use it as a dependency. Like this:

```go
analyzer := NewNodeAnalyzer() // the new struct that would hold the body of collectInventory
scanner = nodeinventorizer.NewCachingScanner(
				analyzer,
				"/cache/inventory-cache",
				env.NodeScanCacheDuration.DurationSetting(),
				env.NodeScanInitialBackoff.DurationSetting(),
				env.NodeScanMaxBackoff.DurationSetting(),
				func(duration time.Duration) { time.Sleep(duration) })
```

Note that both `analyzer` and `scanner` would implement `nodeInventorizer` and passing `scanner` into `scanner` would be theoretically possible ;) 

This would allow us to export away the entire body of the `collectInventory` - note that this function has nothing to do with caching and placing it somewhere else would allow us to make this file more readable.

Reply:
So this would mean simply restoring the deleted `compliance/collection/nodeinventorizer/nodeinventory.go` and adding one more param to the constructor of the caching inventorizer.